File: code/csharp/LnnMoeBook.Examples/LTC/SimpleLtcCell.cs
using System.Globalization;
using LnnMoeBook.Core.Dynamics;
using TorchSharp;

namespace LnnMoeBook.Examples.LTC;

public sealed record LtcSequenceDataset(
    float[] Inputs,
    float[] DeltaTimes,
    float[] Targets,
    int SequenceCount,
    int SequenceLength)
{
    public float InputAt(int sequence, int time) => Inputs[(sequence * SequenceLength) + time];
    public float DeltaTimeAt(int sequence, int time) => DeltaTimes[(sequence * SequenceLength) + time];

    public torch.Tensor ToInputTensor()
    {
        return torch.tensor(Inputs, dtype: torch.float32).reshape(SequenceCount, SequenceLength, 1);
    }

    public torch.Tensor ToDeltaTimeTensor()
    {
        return torch.tensor(DeltaTimes, dtype: torch.float32).reshape(SequenceCount, SequenceLength);
    }

    public torch.Tensor ToTargetTensor()
    {
        return torch.tensor(Targets, dtype: torch.float32).reshape(SequenceCount, 1);
    }
}

public sealed record LtcParameters(
    float InputWeight,
    float RecurrentWeight,
    float GateBias,
    float BaseTimeConstant,
    float Conductance,
    float LeakPotential,
    float ReversalPotential,
    float OutputWeight,
    float OutputBias)
{
    public static LtcParameters Teacher => new(
        InputWeight: 1.10f,
        RecurrentWeight: 0.25f,
        GateBias: -0.10f,
        BaseTimeConstant: 0.70f,
        Conductance: 1.35f,
        LeakPotential: -0.08f,
        ReversalPotential: 0.82f,
        OutputWeight: 1.20f,
        OutputBias: -0.03f);

    public static LtcParameters Student => new(
        InputWeight: 0.78f,
        RecurrentWeight: 0.14f,
        GateBias: -0.02f,
        BaseTimeConstant: 0.86f,
        Conductance: 0.92f,
        LeakPotential: -0.04f,
        ReversalPotential: 0.68f,
        OutputWeight: 0.98f,
        OutputBias: 0.00f);
}

public sealed record LtcSimulationOptions(
    OdeSolverKind SolverKind,
    float InternalStepSize)
{
    public static LtcSimulationOptions Default => new(
        SolverKind: OdeSolverKind.Rk4,
        InternalStepSize: 0.02f);
}

public sealed record LtcTrainingOptions(
    int Iterations,
    float LearningRate,
    float Epsilon)
{
    public static LtcTrainingOptions Default => new(
        Iterations: 60,
        LearningRate: 0.22f,
        Epsilon: 1e-3f);
}

public sealed record LtcStateProperties(
    float Gate,
    float EffectiveTimeConstant,
    float Derivative);

public sealed record LtcStepSnapshot(
    int Sequence,
    int Time,
    float Input,
    float DeltaTime,
    float StateBefore,
    float StateAfter,
    float GateAfter,
    float EffectiveTimeConstantAfter,
    float DerivativeAfter);

public sealed record LtcSequencePrediction(
    float Output,
    float FinalState,
    IReadOnlyList<LtcStepSnapshot> Steps);

public sealed record LtcParameterGradients(
    float InputWeight,
    float RecurrentWeight,
    float GateBias,
    float Conductance,
    float ReversalPotential,
    float OutputWeight,
    float OutputBias);

public sealed record LtcTrainingReport(
    LtcParameters InitialParameters,
    LtcParameters FinalParameters,
    LtcTrainingOptions TrainingOptions,
    IReadOnlyList<float> LossHistory)
{
    public float InitialLoss => LossHistory[0];
    public float FinalLoss => LossHistory[^1];
}

public sealed record LtcSimulationReport(
    LtcParameters Parameters,
    LtcSimulationOptions SimulationOptions,
    LtcTrainingOptions TrainingOptions,
    LtcSequenceDataset Dataset,
    LtcTrainingReport Training,
    IReadOnlyList<float> Predictions,
    float MinEffectiveTimeConstant,
    float MaxEffectiveTimeConstant);

public static class SimpleLtcCell
{
    public static LtcSimulationReport RunDefault()
    {
        var simulationOptions = LtcSimulationOptions.Default;
        var trainingOptions = LtcTrainingOptions.Default;
        var dataset = GenerateSyntheticSequences(
            sequenceCount: 32,
            sequenceLength: 7,
            options: simulationOptions);
        var training = Train(
            LtcParameters.Student,
            dataset,
            simulationOptions,
            trainingOptions);
        var predictions = PredictAll(training.FinalParameters, dataset, simulationOptions);
        var tauRange = EffectiveTimeConstantRange(training.FinalParameters, dataset, simulationOptions);

        return new LtcSimulationReport(
            training.FinalParameters,
            simulationOptions,
            trainingOptions,
            dataset,
            training,
            predictions,
            tauRange.Min,
            tauRange.Max);
    }

    public static LtcSequenceDataset GenerateSyntheticSequences(
        int sequenceCount,
        int sequenceLength,
        LtcSimulationOptions options)
    {
        if (sequenceCount <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(sequenceCount), "Sequence count must be positive.");
        }

        if (sequenceLength <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(sequenceLength), "Sequence length must be positive.");
        }

        ValidateOptions(options);

        var inputs = new float[sequenceCount * sequenceLength];
        var deltaTimes = new float[sequenceCount * sequenceLength];
        var targets = new float[sequenceCount];

        for (var sequence = 0; sequence < sequenceCount; sequence++)
        {
            var phase = sequence * 0.19f;
            for (var time = 0; time < sequenceLength; time++)
            {
                var slow = MathF.Sin(phase + (time * 0.37f));
                var fast = 0.35f * MathF.Sin((sequence * 0.11f) + (time * 0.91f));
                inputs[(sequence * sequenceLength) + time] = slow + fast;
                deltaTimes[(sequence * sequenceLength) + time] = 0.045f + (0.018f * ((sequence + (2 * time)) % 5));
            }

            targets[sequence] = PredictSequence(
                LtcParameters.Teacher,
                inputs,
                deltaTimes,
                sequence,
                sequenceLength,
                options).Output;
        }

        return new LtcSequenceDataset(inputs, deltaTimes, targets, sequenceCount, sequenceLength);
    }

    public static LtcSequencePrediction PredictSequence(
        LtcParameters parameters,
        LtcSequenceDataset dataset,
        int sequence,
        LtcSimulationOptions options)
    {
        ValidateDataset(dataset);

        if (sequence < 0 || sequence >= dataset.SequenceCount)
        {
            throw new ArgumentOutOfRangeException(nameof(sequence), "Sequence index is out of range.");
        }

        return PredictSequence(
            parameters,
            dataset.Inputs,
            dataset.DeltaTimes,
            sequence,
            dataset.SequenceLength,
            options);
    }

    public static IReadOnlyList<float> PredictAll(
        LtcParameters parameters,
        LtcSequenceDataset dataset,
        LtcSimulationOptions options)
    {
        ValidateDataset(dataset);
        ValidateParameters(parameters);
        ValidateOptions(options);

        var predictions = new float[dataset.SequenceCount];
        for (var sequence = 0; sequence < dataset.SequenceCount; sequence++)
        {
            predictions[sequence] = PredictSequence(parameters, dataset, sequence, options).Output;
        }

        return predictions;
    }

    public static float MeanSquaredError(
        LtcParameters parameters,
        LtcSequenceDataset dataset,
        LtcSimulationOptions options)
    {
        var predictions = PredictAll(parameters, dataset, options).ToArray();

        using var predictedTensor = torch.tensor(predictions, dtype: torch.float32);
        using var targetTensor = torch.tensor(dataset.Targets, dtype: torch.float32);
        using var error = predictedTensor - targetTensor;
        using var squared = error * error;
        using var loss = squared.mean();

        return loss.ToSingle();
    }

    public static LtcTrainingReport Train(
        LtcParameters initialParameters,
        LtcSequenceDataset dataset,
        LtcSimulationOptions simulationOptions,
        LtcTrainingOptions trainingOptions)
    {
        ValidateDataset(dataset);
        ValidateParameters(initialParameters);
        ValidateOptions(simulationOptions);
        ValidateTrainingOptions(trainingOptions);

        var parameters = initialParameters;
        var history = new List<float>(trainingOptions.Iterations + 1)
        {
            MeanSquaredError(parameters, dataset, simulationOptions)
        };

        for (var iteration = 0; iteration < trainingOptions.Iterations; iteration++)
        {
            var gradients = EstimateTrainableGradients(
                parameters,
                dataset,
                simulationOptions,
                trainingOptions.Epsilon);

            parameters = ApplyGradients(
                parameters,
                gradients,
                trainingOptions.LearningRate);

            history.Add(MeanSquaredError(parameters, dataset, simulationOptions));
        }

        return new LtcTrainingReport(
            initialParameters,
            parameters,
            trainingOptions,
            history);
    }

    public static LtcParameterGradients EstimateTrainableGradients(
        LtcParameters parameters,
        LtcSequenceDataset dataset,
        LtcSimulationOptions options,
        float epsilon = 1e-3f)
    {
        if (epsilon <= 0.0f)
        {
            throw new ArgumentOutOfRangeException(nameof(epsilon), "Epsilon must be positive.");
        }

        return new LtcParameterGradients(
            EstimateGradient(parameters, dataset, options, epsilon, value => parameters with { InputWeight = value }, parameters.InputWeight),
            EstimateGradient(parameters, dataset, options, epsilon, value => parameters with { RecurrentWeight = value }, parameters.RecurrentWeight),
            EstimateGradient(parameters, dataset, options, epsilon, value => parameters with { GateBias = value }, parameters.GateBias),
            EstimateGradient(parameters, dataset, options, epsilon, value => parameters with { Conductance = MathF.Max(0.02f, value) }, parameters.Conductance),
            EstimateGradient(parameters, dataset, options, epsilon, value => parameters with { ReversalPotential = value }, parameters.ReversalPotential),
            EstimateGradient(parameters, dataset, options, epsilon, value => parameters with { OutputWeight = value }, parameters.OutputWeight),
            EstimateGradient(parameters, dataset, options, epsilon, value => parameters with { OutputBias = value }, parameters.OutputBias));
    }

    public static LtcParameters ApplyGradients(
        LtcParameters parameters,
        LtcParameterGradients gradients,
        float learningRate)
    {
        if (learningRate <= 0.0f)
        {
            throw new ArgumentOutOfRangeException(nameof(learningRate), "Learning rate must be positive.");
        }

        return parameters with
        {
            InputWeight = parameters.InputWeight - (learningRate * gradients.InputWeight),
            RecurrentWeight = parameters.RecurrentWeight - (learningRate * gradients.RecurrentWeight),
            GateBias = parameters.GateBias - (learningRate * gradients.GateBias),
            Conductance = MathF.Max(0.02f, parameters.Conductance - (learningRate * gradients.Conductance)),
            ReversalPotential = parameters.ReversalPotential - (learningRate * gradients.ReversalPotential),
            OutputWeight = parameters.OutputWeight - (learningRate * gradients.OutputWeight),
            OutputBias = parameters.OutputBias - (learningRate * gradients.OutputBias)
        };
    }

    public static LtcStateProperties ComputeStateProperties(
        LtcParameters parameters,
        float input,
        float state)
    {
        ValidateParameters(parameters);

        var gate = Sigmoid(
            (parameters.InputWeight * input)
            + (parameters.RecurrentWeight * state)
            + parameters.GateBias);
        var inverseTau = (1.0f / parameters.BaseTimeConstant) + (parameters.Conductance * gate);
        var effectiveTau = 1.0f / inverseTau;
        var derivative = ((parameters.LeakPotential - state) / parameters.BaseTimeConstant)
            + (parameters.Conductance * gate * (parameters.ReversalPotential - state));

        return new LtcStateProperties(gate, effectiveTau, derivative);
    }

    public static string FormatReport(LtcSimulationReport report)
    {
        return string.Create(
            CultureInfo.InvariantCulture,
            $"ltc cell: sequences={report.Dataset.SequenceCount}, length={report.Dataset.SequenceLength}, solver={report.SimulationOptions.SolverKind}, loss={report.Training.InitialLoss:0.######}->{report.Training.FinalLoss:0.######}, tau=[{report.MinEffectiveTimeConstant:0.######},{report.MaxEffectiveTimeConstant:0.######}]");
    }

    private static LtcSequencePrediction PredictSequence(
        LtcParameters parameters,
        float[] inputs,
        float[] deltaTimes,
        int sequence,
        int sequenceLength,
        LtcSimulationOptions options)
    {
        ValidateParameters(parameters);
        ValidateOptions(options);

        var state = 0.0f;
        var steps = new List<LtcStepSnapshot>(sequenceLength);

        for (var time = 0; time < sequenceLength; time++)
        {
            var input = inputs[(sequence * sequenceLength) + time];
            var deltaTime = deltaTimes[(sequence * sequenceLength) + time];
            var stateBefore = state;
            state = IntegrateState(parameters, input, state, deltaTime, options);
            var properties = ComputeStateProperties(parameters, input, state);

            steps.Add(new LtcStepSnapshot(
                sequence,
                time,
                input,
                deltaTime,
                stateBefore,
                state,
                properties.Gate,
                properties.EffectiveTimeConstant,
                properties.Derivative));
        }

        return new LtcSequencePrediction(
            (parameters.OutputWeight * state) + parameters.OutputBias,
            state,
            steps);
    }

    private static (float Min, float Max) EffectiveTimeConstantRange(
        LtcParameters parameters,
        LtcSequenceDataset dataset,
        LtcSimulationOptions options)
    {
        var min = float.PositiveInfinity;
        var max = float.NegativeInfinity;

        for (var sequence = 0; sequence < dataset.SequenceCount; sequence++)
        {
            var prediction = PredictSequence(parameters, dataset, sequence, options);
            foreach (var step in prediction.Steps)
            {
                min = MathF.Min(min, step.EffectiveTimeConstantAfter);
                max = MathF.Max(max, step.EffectiveTimeConstantAfter);
            }
        }

        return (min, max);
    }

    private static float IntegrateState(
        LtcParameters parameters,
        float input,
        float state,
        float deltaTime,
        LtcSimulationOptions options)
    {
        if (deltaTime <= 0.0f)
        {
            throw new ArgumentOutOfRangeException(nameof(deltaTime), "Delta time must be positive.");
        }

        var subSteps = Math.Max(1, (int)Math.Ceiling(deltaTime / options.InternalStepSize));
        var solver = OdeSolverFactory.Create(options.SolverKind);
        var solution = solver.Solve(new OdeInitialValueProblem(
            Derivative: (_, currentState) => ComputeStateProperties(parameters, input, (float)currentState).Derivative,
            InitialTime: 0.0,
            InitialState: state,
            StepSize: deltaTime / subSteps,
            StepCount: subSteps));

        return (float)solution.FinalPoint.State;
    }

    private static float EstimateGradient(
        LtcParameters parameters,
        LtcSequenceDataset dataset,
        LtcSimulationOptions options,
        float epsilon,
        Func<float, LtcParameters> update,
        float currentValue)
    {
        var plusLoss = MeanSquaredError(update(currentValue + epsilon), dataset, options);
        var minusLoss = MeanSquaredError(update(currentValue - epsilon), dataset, options);

        return (plusLoss - minusLoss) / (2.0f * epsilon);
    }

    private static float Sigmoid(float value)
    {
        var clamped = Math.Clamp(value, -30.0f, 30.0f);
        return 1.0f / (1.0f + MathF.Exp(-clamped));
    }

    private static void ValidateDataset(LtcSequenceDataset dataset)
    {
        if (dataset.SequenceCount <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(dataset), "Dataset must contain at least one sequence.");
        }

        if (dataset.SequenceLength <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(dataset), "Sequence length must be positive.");
        }

        if (dataset.Inputs.Length != dataset.SequenceCount * dataset.SequenceLength)
        {
            throw new ArgumentException("Input array length must be sequenceCount * sequenceLength.", nameof(dataset));
        }

        if (dataset.DeltaTimes.Length != dataset.SequenceCount * dataset.SequenceLength)
        {
            throw new ArgumentException("Delta-time array length must be sequenceCount * sequenceLength.", nameof(dataset));
        }

        if (dataset.Targets.Length != dataset.SequenceCount)
        {
            throw new ArgumentException("Target array length must match sequence count.", nameof(dataset));
        }
    }

    private static void ValidateParameters(LtcParameters parameters)
    {
        if (parameters.BaseTimeConstant <= 0.0f)
        {
            throw new ArgumentOutOfRangeException(nameof(parameters), "Base time constant must be positive.");
        }

        if (parameters.Conductance <= 0.0f)
        {
            throw new ArgumentOutOfRangeException(nameof(parameters), "Conductance must be positive.");
        }
    }

    private static void ValidateOptions(LtcSimulationOptions options)
    {
        if (options.InternalStepSize <= 0.0f)
        {
            throw new ArgumentOutOfRangeException(nameof(options), "Internal step size must be positive.");
        }
    }

    private static void ValidateTrainingOptions(LtcTrainingOptions options)
    {
        if (options.Iterations < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(options), "Iteration count must be non-negative.");
        }

        if (options.LearningRate <= 0.0f)
        {
            throw new ArgumentOutOfRangeException(nameof(options), "Learning rate must be positive.");
        }

        if (options.Epsilon <= 0.0f)
        {
            throw new ArgumentOutOfRangeException(nameof(options), "Epsilon must be positive.");
        }
    }
}

File: code/python/examples/simple_ltc_cell.py
from __future__ import annotations

from dataclasses import dataclass
import math

try:
    import torch
    import torch.nn.functional as F
except ModuleNotFoundError as exc:
    raise SystemExit(
        "PyTorch is required. Install it with: "
        "python -m pip install -r code/python/requirements.txt"
    ) from exc


@dataclass(frozen=True)
class LtcTrainingResult:
    sequence_count: int
    sequence_length: int
    completed_epochs: int
    initial_loss: float
    final_loss: float
    min_effective_tau: float
    max_effective_tau: float


def inverse_softplus(value: float) -> float:
    return math.log(math.exp(value) - 1.0)


class SimpleLtcCell(torch.nn.Module):
    def __init__(
        self,
        *,
        input_weight: float,
        recurrent_weight: float,
        gate_bias: float,
        base_time_constant: float,
        conductance: float,
        leak_potential: float,
        reversal_potential: float,
        output_weight: float,
        output_bias: float,
    ) -> None:
        super().__init__()
        self.input_weight = torch.nn.Parameter(torch.tensor(input_weight, dtype=torch.float32))
        self.recurrent_weight = torch.nn.Parameter(torch.tensor(recurrent_weight, dtype=torch.float32))
        self.gate_bias = torch.nn.Parameter(torch.tensor(gate_bias, dtype=torch.float32))
        self.raw_conductance = torch.nn.Parameter(
            torch.tensor(inverse_softplus(conductance - 0.02), dtype=torch.float32)
        )
        self.reversal_potential = torch.nn.Parameter(torch.tensor(reversal_potential, dtype=torch.float32))
        self.output_weight = torch.nn.Parameter(torch.tensor(output_weight, dtype=torch.float32))
        self.output_bias = torch.nn.Parameter(torch.tensor(output_bias, dtype=torch.float32))
        self.register_buffer("base_time_constant", torch.tensor(base_time_constant, dtype=torch.float32))
        self.register_buffer("leak_potential", torch.tensor(leak_potential, dtype=torch.float32))

    @property
    def conductance(self) -> torch.Tensor:
        return F.softplus(self.raw_conductance) + 0.02

    def state_properties(
        self,
        input_value: torch.Tensor,
        state: torch.Tensor,
    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
        gate = torch.sigmoid(
            (self.input_weight * input_value)
            + (self.recurrent_weight * state)
            + self.gate_bias
        )
        inverse_tau = (1.0 / self.base_time_constant) + (self.conductance * gate)
        effective_tau = 1.0 / inverse_tau
        derivative = ((self.leak_potential - state) / self.base_time_constant) + (
            self.conductance * gate * (self.reversal_potential - state)
        )
        return gate, effective_tau, derivative

    def forward(self, inputs: torch.Tensor, delta_times: torch.Tensor) -> torch.Tensor:
        state = torch.zeros(inputs.shape[0], 1, dtype=inputs.dtype, device=inputs.device)
        for time in range(inputs.shape[1]):
            input_value = inputs[:, time, :]
            delta_time = delta_times[:, time].unsqueeze(-1)
            _, _, derivative = self.state_properties(input_value, state)
            state = state + (delta_time * derivative)

        return (self.output_weight * state) + self.output_bias

    def effective_tau_range(
        self,
        inputs: torch.Tensor,
        delta_times: torch.Tensor,
    ) -> tuple[float, float]:
        state = torch.zeros(inputs.shape[0], 1, dtype=inputs.dtype, device=inputs.device)
        values: list[torch.Tensor] = []

        with torch.no_grad():
            for time in range(inputs.shape[1]):
                input_value = inputs[:, time, :]
                delta_time = delta_times[:, time].unsqueeze(-1)
                _, tau, derivative = self.state_properties(input_value, state)
                values.append(tau)
                state = state + (delta_time * derivative)

            all_tau = torch.cat(values, dim=0)
            return float(all_tau.min()), float(all_tau.max())


def teacher_cell() -> SimpleLtcCell:
    return SimpleLtcCell(
        input_weight=1.10,
        recurrent_weight=0.25,
        gate_bias=-0.10,
        base_time_constant=0.70,
        conductance=1.35,
        leak_potential=-0.08,
        reversal_potential=0.82,
        output_weight=1.20,
        output_bias=-0.03,
    )


def student_cell() -> SimpleLtcCell:
    return SimpleLtcCell(
        input_weight=0.78,
        recurrent_weight=0.14,
        gate_bias=-0.02,
        base_time_constant=0.86,
        conductance=0.92,
        leak_potential=-0.04,
        reversal_potential=0.68,
        output_weight=0.98,
        output_bias=0.00,
    )


def generate_synthetic_sequences(
    sequence_count: int = 32,
    sequence_length: int = 7,
) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
    if sequence_count <= 0:
        raise ValueError("sequence_count must be positive")
    if sequence_length <= 0:
        raise ValueError("sequence_length must be positive")

    inputs = torch.zeros(sequence_count, sequence_length, 1, dtype=torch.float32)
    delta_times = torch.zeros(sequence_count, sequence_length, dtype=torch.float32)

    for sequence in range(sequence_count):
        phase = sequence * 0.19
        for time in range(sequence_length):
            slow = math.sin(phase + (time * 0.37))
            fast = 0.35 * math.sin((sequence * 0.11) + (time * 0.91))
            inputs[sequence, time, 0] = slow + fast
            delta_times[sequence, time] = 0.045 + (0.018 * ((sequence + (2 * time)) % 5))

    teacher = teacher_cell()
    with torch.no_grad():
        targets = teacher(inputs, delta_times)

    return inputs, delta_times, targets


def train(epochs: int = 200, learning_rate: float = 0.05) -> LtcTrainingResult:
    if epochs <= 0:
        raise ValueError("epochs must be positive")
    if learning_rate <= 0:
        raise ValueError("learning_rate must be positive")

    inputs, delta_times, targets = generate_synthetic_sequences()
    model = student_cell()
    optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
    losses: list[float] = []

    for _ in range(epochs + 1):
        prediction = model(inputs, delta_times)
        loss = torch.mean((prediction - targets) ** 2)
        losses.append(float(loss.detach()))

        if len(losses) == epochs + 1:
            break

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

    tau_min, tau_max = model.effective_tau_range(inputs, delta_times)
    return LtcTrainingResult(
        sequence_count=inputs.shape[0],
        sequence_length=inputs.shape[1],
        completed_epochs=epochs,
        initial_loss=losses[0],
        final_loss=losses[-1],
        min_effective_tau=tau_min,
        max_effective_tau=tau_max,
    )


def format_report(result: LtcTrainingResult) -> str:
    return (
        "ltc cell: "
        f"sequences={result.sequence_count}, "
        f"length={result.sequence_length}, "
        f"epochs={result.completed_epochs}, "
        f"loss={result.initial_loss:.6f}->{result.final_loss:.6f}, "
        f"tau=[{result.min_effective_tau:.6f},{result.max_effective_tau:.6f}]"
    )


if __name__ == "__main__":
    print(format_report(train()))
